// @salesforce/core/lib/sfdxError.js
"use strict";
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const kit_1 = require("@salesforce/kit");
const ts_types_1 = require("@salesforce/ts-types");
const messages_1 = require("./messages");
/**
 * A class to manage all the keys and tokens for a message bundle to use with SfdxError.
 *
 * @example
 * SfdxError.create(new SfdxErrorConfig('MyPackage', 'apex', 'runTest').addAction('apexErrorAction1', [className]));
 */
class SfdxErrorConfig {
    /**
     * Create a new SfdxErrorConfig.
     * @param {string} packageName The name of the package.
     * @param {string} bundleName The message bundle.
     * @param {string} errorKey The error message key.
     * @param {Tokens} errorTokens The tokens to use when getting the error message.
     * @param {string} [actionKey] The action message key.
     * @param {Tokens} [actionTokens] The tokens to use when getting the action message(s).
     */
    constructor(packageName, bundleName, errorKey, errorTokens = [], actionKey, actionTokens) {
        this.actions = new Map();
        this.packageName = packageName;
        this.bundleName = bundleName;
        this.errorKey = errorKey;
        this.errorTokens = errorTokens;
        if (actionKey)
            this.addAction(actionKey, actionTokens);
    }
    /**
     * Set the error key.
     * @param {string} key The key to set.
     * @returns {SfdxErrorConfig} For convenience `this` object is returned.
     */
    setErrorKey(key) {
        this.errorKey = key;
        return this;
    }
    /**
     * Set the error tokens.
     * @param {Tokens} tokens The tokens to set.
     * @returns {SfdxErrorConfig} For convenience `this` object is returned.
     */
    setErrorTokens(tokens) {
        this.errorTokens = tokens;
        return this;
    }
    /**
     * Add an error action to assist the user with a resolution.
     * @param {string} actionKey The action key in the message bundle.
     * @param {Tokens} [actionTokens] The action tokens for the string.
     * @returns {SfdxErrorConfig} For convenience `this` object is returned.
     */
    addAction(actionKey, actionTokens = []) {
        this.actions.set(actionKey, actionTokens);
        return this;
    }
    /**
     * Load the messages using Messages.loadMessages.
     * @returns {Messages} Returns the loaded messages.
     */
    load() {
        this.messages = messages_1.Messages.loadMessages(this.packageName, this.bundleName);
        return this.messages;
    }
    /**
     * Get the error message using messages.getMessage.
     * @returns {string}
     * @throws {SfdxError} If errorMessages.load was not called first.
     */
    getError() {
        if (!this.messages) {
            throw new SfdxError('SfdxErrorConfig not loaded.');
        }
        return this.messages.getMessage(this.errorKey, this.errorTokens);
    }
    /**
     * Get the action messages using messages.getMessage.
     * @returns {Optional<string[]>} List of action messages.
     * @throws {SfdxError} If errorMessages.load was not called first.
     */
    getActions() {
        if (!this.messages) {
            throw new SfdxError('SfdxErrorConfig not loaded.');
        }
        if (this.actions.size === 0)
            return;
        const actions = [];
        this.actions.forEach((tokens, key) => {
            const messages = this.messages;
            if (messages) {
                actions.push(messages.getMessage(key, tokens));
            }
        });
        return actions;
    }
    /**
     * Remove all actions from this error config. Useful when reusing SfdxErrorConfig.
     * for other error messages within the same bundle.
     * @returns {SfdxErrorConfig} For convenience `this` object is returned.
     */
    removeActions() {
        this.actions = new Map();
        return this;
    }
}
exports.SfdxErrorConfig = SfdxErrorConfig;
/**
 * A generalized sfdx error which also contains an action. The action is used in the
 * CLI to help guide users past the error.
 *
 * To throw an error in a synchronous function you must either pass the error message and actions
 * directly to the constructor, e.g.
 *
 * @example
 * // To load a message bundle:
 * Messages.importMessagesDirectory(__dirname);
 * this.messages = Messages.loadMessages('myPackageName', 'myBundleName');
 * // Note that __dirname should contain a messages folder.
 *
 * // To throw an error associated with the message from the bundle:
 * throw SfdxError.create('myPackageName', 'myBundleName', 'MyErrorMessageKey', [messageToken1]);
 *
 * // To throw a non-bundle based error:
 * throw new SfdxError(myErrMsg, 'MyErrorName');
 *
 */
class SfdxError extends kit_1.NamedError {
    // The create implementation function.
    static create(nameOrConfig, bundleName, key, tokens) {
        let errorConfig;
        if (ts_types_1.isString(nameOrConfig)) {
            errorConfig = new SfdxErrorConfig(nameOrConfig, ts_types_1.ensure(bundleName), ts_types_1.ensure(key), tokens);
        }
        else {
            errorConfig = nameOrConfig;
        }
        errorConfig.load();
        return new SfdxError(errorConfig.getError(), errorConfig.errorKey, errorConfig.getActions());
    }
    /**
     * Convert an Error to an SfdxError.
     * @param {Error} err The error to convert.
     * @returns {SfdxError}
     */
    static wrap(err) {
        const sfdxError = new SfdxError(err.message, err.name);
        if (sfdxError.stack) {
            sfdxError.stack = sfdxError.stack.replace(`${err.name}: ${err.message}`, 'Outer stack:');
            sfdxError.stack = `${err.stack}\n${sfdxError.stack}`;
        }
        return sfdxError;
    }
    /**
     * Create an SfdxError.
     * @param {string} message The error message.
     * @param {string} [name] The error name. Defaults to 'SfdxError'.
     * @param {string[]} [actions] The action message(s).
     * @param {number} [exitCode] The exit code which will be used by SfdxCommand.
     * @param {Error} [cause] The underlying error that caused this error to be raised.
     */
    constructor(message, name, actions, exitCode, cause) {
        super(name || 'SfdxError', message, cause);
        this.actions = actions;
        this.exitCode = exitCode || 1;
    }
    /**
     * Sets the name of the command.
     * @param {string} commandName The command name.
     * @returns {SfdxError} For convenience `this` object is returned.
     */
    setCommandName(commandName) {
        this.commandName = commandName;
        return this;
    }
    /**
     * An additional payload for the error.
     * @param {any} data The payload data.
     * @returns {SfdxError} For convenience `this` object is returned.
     */
    setData(data) {
        this.data = data;
        return this;
    }
    /**
     * Convert an SfdxError's state to an object.
     * @returns {object} A plain object representing the state of this error.
     */
    toObject() {
        const obj = {
            name: this.name,
            message: this.message || this.name,
            exitCode: this.exitCode,
            actions: this.actions
        };
        if (this.commandName) {
            obj.commandName = this.commandName;
        }
        if (this.data) {
            obj.data = this.data;
        }
        return obj;
    }
}
exports.SfdxError = SfdxError;
