// @salesforce/core/lib/config/orgUsersConfig.js
"use strict";
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const configFile_1 = require("./configFile");
/**
 * A config file that stores usernames for an org.
 *
 * @extends ConfigFile
 */
class OrgUsersConfig extends configFile_1.ConfigFile {
    /**
     * **Do not use.** Call {@link OrgUsersConfig.getOptions} instead.
     */
    static getDefaultOptions(isGlobal, filename) {
        throw Error('The method OrgUsersConfig.getDefaultOptions is not supported. Call OrgUsersConfig.getOptions().');
    }
    /**
     * Gets the config options for a given org ID.
     * @param {string} orgId The orgId. Generally this org would have multiple users configured.
     * @return {ConfigOptions} The ConfigOptions.
     */
    static getOptions(orgId) {
        return {
            isGlobal: true,
            isState: true,
            filename: `${orgId}.json`
        };
    }
}
exports.OrgUsersConfig = OrgUsersConfig;
