// @salesforce/core/lib/util/time.js
"use strict";
/*
 * Copyright (c) 2016, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Units of time
 */
var TIME_UNIT;
(function (TIME_UNIT) {
    TIME_UNIT[TIME_UNIT["MINUTES"] = 0] = "MINUTES";
    TIME_UNIT[TIME_UNIT["MILLISECONDS"] = 1] = "MILLISECONDS";
    TIME_UNIT[TIME_UNIT["SECONDS"] = 2] = "SECONDS";
})(TIME_UNIT = exports.TIME_UNIT || (exports.TIME_UNIT = {}));
/**
 * Helper class to convert cli times between minutes, seconds, and milliseconds.
 */
class Time {
    constructor(quantity, unit = TIME_UNIT.MINUTES) {
        this.quantity = quantity;
        this.unit = unit;
    }
    get minutes() {
        switch (this.unit) {
            case TIME_UNIT.MINUTES:
                return this.unit;
            case TIME_UNIT.SECONDS:
                return Math.round(this.quantity / Time.SECONDS_IN_MINUTE);
            case TIME_UNIT.MILLISECONDS:
                return Math.round(this.quantity / Time.MILLI_IN_SECONDS / Time.SECONDS_IN_MINUTE);
            default:
                return -1;
        }
    }
    get seconds() {
        switch (this.unit) {
            case TIME_UNIT.MINUTES:
                return this.quantity * Time.SECONDS_IN_MINUTE;
            case TIME_UNIT.SECONDS:
                return this.quantity;
            case TIME_UNIT.MILLISECONDS:
                return Math.round(this.quantity / Time.MILLI_IN_SECONDS);
            default:
                return -1;
        }
    }
    get milliseconds() {
        switch (this.unit) {
            case TIME_UNIT.MINUTES:
                return this.quantity * Time.SECONDS_IN_MINUTE * Time.MILLI_IN_SECONDS;
            case TIME_UNIT.SECONDS:
                return this.quantity * Time.MILLI_IN_SECONDS;
            case TIME_UNIT.MILLISECONDS:
                return this.quantity;
            default:
                return -1;
        }
    }
}
Time.MILLI_IN_SECONDS = 1000;
Time.SECONDS_IN_MINUTE = 60;
exports.Time = Time;
